### PR TITLE
Adds 'runAsUser' service account permission checking as prerequisite …

### PR DESCRIPTION
**Pull request by ttomsu**
…to saving a pipeline.

Tests won't pass until https://github.com/spinnaker/fiat/pull/108 is merged and released.

tag: https://github.com/spinnaker/fiat/issues/36

### CODE REVIEW

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/AuthorizationSupport.groovy [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.stereotype.Component
+
+@Component
+class AuthorizationSupport {
+
+  @Autowired
+  FiatPermissionEvaluator permissionEvaluator
+
+  boolean hasRunAsUserPermission(Pipeline pipeline) {
+    List<String> runAsUsers = pipeline.triggers*.runAsUser
```
Comment (jtk54) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/AuthorizationSupport.groovy]: Not all pipelines have `triggers`, right? Might need a `?` and/or a `?: []`.

Comment (ttomsu) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/AuthorizationSupport.groovy]: My testing of the spread operator shows it to be null safe.

https://groovy-playground.appspot.com/#?load=8493a5b4e9ebea349d4d99705e964ef2

**Comment by ajordens** [comment]
So you're basically enforcing that if we're attempting to save a pipeline w/ a `runAsUser` trigger that the authenticated user actually has access to the trigger user?

Do strategies need similar treatment?

**Comment by ajordens** [comment]
Sorry, didn't mean to close.

**Comment by ttomsu** [comment]
Your understanding is correct. We have to do this on pipeline save so that Echo doesn't have to do (and handle the failure of) an auth check.

A cursory look at strategy config doesn't show an ability to trigger a strategy outside of a "deploy" stage. So no, I don't think you'd be able to alter the trigger user from a strategy.

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy [hunk 2]
```diff
@@ -46,10 +46,10 @@ class PipelineController {
   @Autowired
   PipelineDAO pipelineDAO
 
-  @PreAuthorize("@fiatPermissionEvaluator.storeWholePermission()")
-  @PostFilter("hasPermission(filterObject.application, 'APPLICATION', 'READ')")
+  @PreAuthorize("#restricted ? @fiatPermissionEvaluator.storeWholePermission() : true")
+  @PostFilter("#restricted ? hasPermission(filterObject.name, 'APPLICATION', 'READ') : true")
   @RequestMapping(value = '', method = RequestMethod.GET)
-  List<Pipeline> list() {
+  List<Pipeline> list(@RequestParam(required = false, value = 'restricted', defaultValue = 'true') boolean restricted) {
```
Comment (ajordens) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy]: Thoughts on this as an HTTP header vs query parameter? Former might align better given that everything else auth-related flows over headers?

Comment (ttomsu) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy]: Either one would work. I opted for a query param for two reasons:

1.) the ?restricted=false appears in the logs by default
2.) consistent with the same thing listing Applications in Clouddriver.

I plan to submit as is, and can file a bug to move both servers and clients over to headers if you want.

**Pull request merged**